blockchain/merkle.go

```diff
@@ -141,3 +140,91 @@ func BuildMerkleTreeStore(transactions []*btcutil.Tx, witness bool) []*wire.ShaH
 
 	return merkles
 }
+
+// ExtractWitnessCommitment attempts to locate, and return the witness
+// commitment for a block. The witness commitment is of the form:
+// SHA256(witness root || witness nonce).The function additionally returns a
+// boolean indicating if the witness root was located within any of the txOut's
+// in the passed transaction. The witness commitment is stored as the data push
+// for an OP_RETURN with special magic bytes to aide in location.
+func ExtractWitnessCommitment(tx *btcutil.Tx) ([]byte, bool) {
+	var witnessCommitment []byte
+
+	// The witness commitment *must* be located within one of the coinbase
+	// transaction's outputs.
+	if !IsCoinBase(tx) {
+		return witnessCommitment, false
+	}
+
+	msgTx := tx.MsgTx()
+	witFound := false
+	for i := len(msgTx.TxOut) - 1; i >= 0; i-- {
+		// The public key script that contains the witness commitment
+		// must shared a prefix with the WitnessMagicBytes, and be at
+		// least 38 bytes.
+		if bytes.HasPrefix(msgTx.TxOut[i].PkScript, WitnessMagicBytes) &&
+			len(msgTx.TxOut[i].PkScript) >= 38 {
+
+			// The witness commitment itself is a 32-byte hash
+			// directly after the WitnessMagicBytes. The remaining
+			// bytes beyond the 38th byte currently have no consensus
+			// meaning.
+			witnessCommitment = msgTx.TxOut[i].PkScript[6:38]
+			witFound = true
+		}
+	}
+
+	return witnessCommitment, witFound
+}
+
+// ValidateWitnessCommitment validates the witness commitment (if any) found
+// within the coinbase transaction of the passed block.
+func ValidateWitnessCommitment(blk *btcutil.Block) error {
+	// TODO(roasbeef): check should be conditional on witness activation
+	coinbaseTx := blk.Transactions()[0]
+	witnessCommitment, witnessFound := ExtractWitnessCommitment(coinbaseTx)
+
+	// If we can't find a witness commitment in any of the coinbase's
+	// outputs, then the block MUST NOT contain any transactions with
+	// witness data.
+	if !witnessFound {
+		for _, tx := range blk.Transactions() {
+			msgTx := tx.MsgTx()
+			if len(msgTx.TxIn[0].Witness) != 0 {
+				str := fmt.Sprintf("block contains transaction with witness" +
+					" data, yet not witness commitment present")
+				return ruleError(ErrUnexpectedWitness, str)
+			}
+		}
+		return nil
+	}
+
+	// If the witness commitment can be located, then the coinbase
+	// transaction MUST have exactly one witness element within its witness
+	// data. Additionally, that element must be exactly 32-byte.
+	if len(coinbaseTx.MsgTx().TxIn[0].Witness) != 1 ||
+		len(coinbaseTx.MsgTx().TxIn[0].Witness[0]) != 32 {
+		str := fmt.Sprintf("the coinbase transaction must have exactly " +
+			"one item within its witness stack, and that item " +
+			"should be exactly 32-bytes")
```

Comment:
Please separate the error messages here so that they contain more details such as:

```Go
	// At this point the block contains a witness commitment, so the
	// coinbase transaction MUST have exactly one witness element within
	// its witness data and that element must be exactly
	// CoinbaseWitnessDataLen bytes.
	coinbaseWitness := coinbaseTx.MsgTx().TxIn[0].Witness
	if len(coinbaseWitness) != 1 {
		str := fmt.Sprintf("the coinbase transaction has %d items in "+
			"its witness stack when 1 was expected",
			len(coinbaseWitness))
		return ruleError(ErrInvalidWitnessCommitment, str)
	}
	if len(coinbaseWitness[0]) != CoinbaseWitnessDataLen {
		str := fmt.Sprintf("the coinbase transaction witness entry "+
			"has %d bytes when %d bytes were expected",
			len(coinbaseWitness[0]), CoinbaseWitnessDataLen)
		return ruleError(ErrInvalidWitnessCommitment, str)
	}
```

Also notice I introduced a constant for that "32" magic.